flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/UpstreamRecoveryTracker.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.runtime.checkpoint.channel.InputChannelInfo;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.HashSet;
+
+final class UpstreamRecoveryTracker {
```

Comment:
We should generalize it for all epochs. Not sure if we already do it in this PR or then later.

Reply:
I'd prefer to do it later if needed.
This PR is already a bit disconnected from how it will be used (tearing down VCs), and the code will probably change.